indexer/SymbolFormatter.h

```diff
@@ -24,11 +26,41 @@ class NamespaceDecl;
 class TagDecl;
 } // namespace clang
 
+namespace llvm {
+class raw_ostream;
+}
+
 namespace scip_clang {
 
 using GetCanonicalPath =
     absl::FunctionRef<std::optional<RootRelativePathRef>(clang::FileID)>;
 
+/// Type similar to \c scip::Descriptor but carrying \c string_view fields
+/// instead to avoid redundant intermediate allocations.
+struct DescriptorBuilder {
+  std::string_view name;
+  std::string_view disambiguator;
+  scip::Descriptor::Suffix suffix;
+
+  void formatTo(llvm::raw_ostream &) const;
+};
+
+/// Type similar to \c scip::Symbol but carrying \c string_view fields instead
+/// to avoid redundant allocations.
+///
+/// Meant for transient use in constructing symbol strings, since a
+/// \c scip::Index doesn't store any \c scip::Symbol values directly.
+struct SymbolBuilder {
+  std::string_view packageName;
+  std::string_view packageVersion;
+  llvm::SmallVector<DescriptorBuilder, 4> descriptors;
+
+  void formatTo(std::string &) const;
+
+  static std::string formatContextual(std::string_view contextSymbol,
```

Comment:
Briefly comment what "contextual" refers to.

Reply:
Added a doc comment, does that seem reasonable?